doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+  (`user`)
+
+### Permissions management for application-specific passwords
+
+New ACIs must automatically be added and maintained for every
+application-specific password so that the owning user will have read-write
+access to all (and only) objects holding his application-specific passwords - 
+both when logged in using the primary password and when logged in using the
+respective application-specific password. This is the only thing to do with
+respect to RBAC and application-specific passwords, as RBAC serves only for work
+with LDAP in FreeIPA.
+
+The only general limitation for subaccounts will be the inability to manage the
+account (modify the information about the user, change his other passwords...).
+As all permissions for them will be assigned by administrators separately and
+from scratch, even a limitation such as reducing the effective permissions of
+the current user session (possibly according to the application policy) using a
+389-DS plugin probably will not be needed.
+
+#### HBAC for application-specific passwords
+
+Two-component principals for subaccounts, as described above, will be used for
+assigning HBAC rules to users logged in using application-specific passwords.
+These HBAC rules will be assigned by administrators just as the other ones.
+
+TBD: When an application-specific password is deleted, delete the associated
+HBAC rules automatically, or keep them?
+
+Optionally, a possibility for administrators to specify that users can log into
+a particular service only using the passwords from a specified LDAP subtree (or
+the primary password) can be provided.
+
+hbacrule plugin therefore will need to be extended with the ability to assign
+allowed LDAP subtrees to application-specific passwords as a part of HBAC rules.
+hbactest plugin will be extended to test this, and it will be used to test HBAC
+with application-specific passwords in general.
+
+### SSSD support for the aforementioned functionality
+
+SSSD will also be extended to support the aforementioned functionality. This
+will be done by a closely related, but separate extension of SSSD. On the
+side of FreeIPA, after SSSD is extended, interactions with SSSD using the new
+functionality must be tested in `ipatests/test_integration/test_sssd.py`.
+
+## Implementation
+
+The extension must be backward compatible, that means not breaking any existing
+functionality and passing all existing tests.
+
+As for changes to the data storage:
+* As described above, a new LDAP subtree will be created for
+  application-specific passwords, with objects holding them being grouped by
+  application, and their attributes being, besides an application-specific
+  password itself, a two-component Kerberos principal containing the application
+  name, and a backlink to the owning user.
+* New ACIs will be automatically created and maintained so that users will have
+  read-write access to objects holding their own application-specific passwords.
```

Comment:
See above, no dynamic ACIs, please, this is important.

Reply:
[All right](https://github.com/freeipa/freeipa/pull/4061#discussion_r400096672).